app/models/topic.rb

```diff
@@ -144,6 +144,15 @@ def public
     forum_id >= 3 && !private?
   end
 
+  def create_topic_with_user(params, current_user)
+    self.user = current_user ? current_user : User.find_by_email(params[:topic][:user][:email]) 
+    
+    unless self.user #User not found, lets build it
+      self.build_user(params[:topic].require(:user).permit(:email, :name)).signup_guest
+    end
+    self.user.persisted? and self.save
+  end  
+
```

Comment:
Nice refactor!  Can you add a spec or two?  I know this is probably covered by the controller tests, but it would be nice to have a model test as well.